Campaign Feature

jkobjacob opened the pull request.

jacobstakeall committed:
Basic campaign contract.

cryptostaker2021 commented on contracts/campaign/Campaign.sol:
```diff
@@ -0,0 +1,70 @@
+        uint256 _totalRewardAmount,
+        IERC20 _rewardToken
+    ) external onlyRole(GOVERNANCE_ROLE) {
+        uint256 totalShuttles_ = _endShuttle.sub(_startShuttle);
```
are these shuttle number inclusive? 
If I want to run for shuttle 20-40, what arguments I will pass? 
19 -40 or 20-40 ?

jacobstakeall replied:
https://github.com/stakeall/cross-chain-staking-shuttle/blob/08f818f2cf07a0aecfc21dec9b46bc1c8a5d831f/contracts/campaign/Campaign.sol#L70

Shuttle Numbers are inclusive.

cryptostaker2021 commented on contracts/campaign/Campaign.sol:
```diff
@@ -0,0 +1,70 @@
+        uint256 _userShare,
+        address payable _sender
+    ) external {
+        uint256 rewardAmount_ = campaigns[_campaignNumber].rewardAmountPerShuttle.mul(_userShare);
```
Add a modifier that only child pool can call this

jacobstakeall replied:
https://github.com/stakeall/cross-chain-staking-shuttle/blob/08f818f2cf07a0aecfc21dec9b46bc1c8a5d831f/contracts/campaign/Campaign.sol#L69

Added.

cryptostaker2021 commented on contracts/campaign/Campaign.sol:
```diff
@@ -0,0 +1,70 @@
+
+    mapping(uint256 => ACampaign) public campaigns;
+
+    function initialize(
```
Accept child pool address

jacobstakeall replied:
https://github.com/stakeall/cross-chain-staking-shuttle/blob/08f818f2cf07a0aecfc21dec9b46bc1c8a5d831f/contracts/campaign/Campaign.sol#L26

Passed childPool to the initialize

cryptostaker2021 commented on contracts/campaign/Campaign.sol:
```diff
@@ -0,0 +1,70 @@
+        uint256 _userShare,
+        address payable _sender
+    ) external {
+        uint256 rewardAmount_ = campaigns[_campaignNumber].rewardAmountPerShuttle.mul(_userShare);
```
This seems wrong. 
It should be rewardAmountPerShuttle * userShare / totalShare

jacobstakeall replied:
https://github.com/stakeall/cross-chain-staking-shuttle/commit/3f3f66e99d67a9cda48761d300260a5172aea248

Fixed.

cryptostaker2021 commented on contracts/campaign/Campaign.sol:
```diff
@@ -0,0 +1,70 @@
+
+    }
+
+    function createCampaign(
```
We need a method to delete a campaign?

jacobstakeall replied:
This requires a discussion.

jacobstakeall committed:
Changes -
- ChildPool address initialized.
- Added modifier onlyChildPool.
- Claim only called by childPool.

jacobstakeall committed:
Require statement to validate shuttle part of campaign.

jacobstakeall committed:
Changes -
- Added extra param "totalAmount" to claim method.
- Fixed the rewardAmount_ math.

jkobjacob commented on contracts/campaign/Campaign.sol:
```diff
@@ -67,6 +67,12 @@ contract Campaign is ICampaign, CampaignSecurityModule {
         uint256 _userShare,
         address payable _sender
     ) external onlyChildPool {
+        require(
```
@cryptostaker2021 The shuttle numbers are inclusive.

jacobstakeall committed:
Fix total shuttle count.

jacobstakeall committed:
Added CampaignStatus.

jacobstakeall committed:
Add methods to modify Campaign status.

jacobstakeall committed:
Changes -
- Rename claim to claimRewards.
- Require statement to allow claim only from campaign with active campaignStatus.

jacobstakeall committed:
Changes ChildPool -
- Added ICampaign and exposed setter method.

jacobstakeall committed:
Expose a new method "ClaimWithRewards" -
- To claim stMatic and rewards if any.

jacobstakeall committed:
Fixes Campaign - args, syntax.

jacobstakeall committed:
Changes ChildPool -
- Refactor "claim()" & "claimRewards()" to use internal method "_claim()".
- totalAmount access from shuttles mapping.

jacobstakeall committed:
Add event for campaignStatusModified.

cryptostaker2021 commented on contracts/campaign/Campaign.sol:
```diff
@@ -0,0 +1,124 @@
+            "!No Shuttle in Campaign"
+        );
+        
+        uint256 rewardAmount_ = campaigns[_campaignNumber].rewardAmountPerShuttle.mul(_userAmount).div(_totalAmount);
```
You are not updating the claimed amount of that campaign? `totalClaimedAmount`

jacobstakeall committed:
Changes Childpool -
- Added documentation.
- Rewards can be claimed only if shuttle is arrived.

jacobstakeall committed:
Changes Campaign -
- Fix require messages.
- Update claimed reward amount for each claim.

jacobstakeall committed:
Test scripts
- upgrade campaign
- deploy campaign

jacobstakeall committed:
Add parameter rewardAmountPerShuttle.

jacobstakeall committed:
Add documentation - Campaign.

jacobstakeall committed:
Add method to unpause a campaign.

jacobstakeall committed:
Fix ChildPool - storage collision.

jacobstakeall committed:
Fixes - syntax and typo errors.

jacobstakeall committed:
Add method "transferUnusedRewards" -
- To pull out rewards which are allocated for a campaign and left unused.
- An example of unused case would be the shuttle part of that campaign got cancelled
provided, the rewards have to be distributed only for successful processing of a shuttle.

jacobstakeall committed:
Added dev documentation.

jacobstakeall committed:
Fix: Declared ICampaign as interface.

jacobstakeall committed:
Make use of cached currentCampaign_ rather accessing storage.

jacobstakeall committed:
Changes includes -
- Replace require with custom errors.
- Cache campaign in memory in claimRewards function.

jacobstakeall committed:
Removed SafeMath library usage.

jacobstakeall committed:
Replaced Storage variable with function parameter in event arg.

jacobstakeall committed:
Extract modifier governance role to an internal function.

jacobstakeall committed:
[ChildPool.sol]: Replace require with Custom Reverts.

jacobstakeall committed:
Fix: If guard -
- shuttleNumber should lie within the closed interval (startShuttle, endShuttle).

jacobstakeall committed:
ChildPool: Revert Custom errors.

jacobstakeall committed:
Replace require with revert - setCampaign().

jacobstakeall committed:
Audit Report: Campaign Feature by Shard Labs.
